documentation/prometheus-design-principles.md

```diff
@@ -0,0 +1,164 @@
+
+### Values
+
+  * [counters should always count up from zero](https://www.robustperception.io/how-does-a-prometheus-counter-work/)
+      * data should always make sense, independent of scrape interval
+      * this is in contrast to statsd's behaviour
+  * [sum of rates, not rate of sums](https://www.robustperception.io/rate-then-sum-never-sum-then-rate/)
+
+## Architectural things ##
+
+### Simplicity ###
+
+Prometheus is designed to achieve reliability through simplicity.
+
+This means:
+  * redundancy is achieved by deploying multiple separate instances,
+    not by database replication.
+  * recovery from backup is not a priority.  if a prometheus server
+    has a disk failure, just use another redundant instance (assuming
+    you have one)
+  * if a prometheus instance has a temporary blip (say, a network
+    issue) and then recovers, it will have missing metrics for the
+    duration of the issue.
+    
+[Julius Volz talked about why they didn't do clustering](https://youtu.be/QgJbxCWRZ1s?t=26m40s).
+
+### Prometheus should be deployed in same failure domain ###
+
+A prometheus server should be deployed in the same failure domain as
+the things it monitors.  This means the same data centre, region,
+availability zone, etc.
+
+This means that:
+
+  * a prometheus shouldn't have to go over the public internet to
+    scrape a service
+      * (federation might go over public internet, though)
+      
+https://www.robustperception.io/scaling-and-federating-prometheus/
+
+### There should normally be a single alertmanager for an organization ###
+
+It seems that there should normally be a single alertmanager for an
+organization.  The best reference I've found for this is from
+[Julius Volz's Prometheus Design and Philosophy talk (7:44)](https://youtu.be/QgJbxCWRZ1s?t=7m43s)
+which doesn't go into much detail as to *why* you'd do this.
+
+## Security model
+
+Prometheus has extensive
+[documentation about its security model](https://prometheus.io/docs/operating/security/).
+In particular, it distinguishes trusted users (who have admin access
+to deploy and reconfigure prometheus) from untrusted users (who can
+view prometheus metrics and dashboards).
+
+Mostly, you control access to prometheus at the network level; there
+is no support for authentication within prometheus.
+
+## Federation
+
+Prometheus supports federation: a prometheus server can scrape the
+metrics from another prometheus server.  This is to support some use
+cases, such as:
+
+  * have a global, coarse-grain view of the system
+      * the global prometheus may have a coarser scrape resolution
+      * the global prometheus may aggregate instance-level time series
+        into service-level
+
+## Alerting on symptoms, not causes
+
+As mentioned above, Prometheus only deals with aggregated numerical
+data, not raw event data.  Prometheus data is used to trigger alerts,
+based on symptoms ("the system is broken") rather than causes.
```

Comment:
"the system is broken" or "the system is about to be broken" e.g. at current rate of disk usage, the disk will fill in 24 hours

Reply:
This type of alerting also highlights Prometheus preference for whitebox instrumentation rather than purely blackbox monitoring you get from something like Pingdom

Reply:
i think this discussion deserves its own section rather than being discussed in parentheses right here in the doc